core/src/main/java/org/infinispan/container/entries/AbstractInternalCacheEntry.java

```diff
@@ -133,4 +172,39 @@ public final boolean equals(Object o) {
    public final int hashCode() {
       return  31 * Objects.hashCode(getKey()) + Objects.hashCode(getValue());
    }
+
+   protected abstract InternalCacheValue createCacheValue();
+
+   protected void appendFieldsToString(StringBuilder builder) {
+      builder.append("key=").append(Util.toStr(key));
+      builder.append(", value=").append(Util.toStr(value));
+      builder.append(", internalMetadata=").append(internalMetadata);
+   }
+
+   protected static void writeCommonDataTo(AbstractInternalCacheEntry cacheEntry, ObjectOutput output)
+         throws IOException {
+      output.writeObject(cacheEntry.key);
+      output.writeObject(cacheEntry.value);
+      output.writeObject(cacheEntry.internalMetadata);
+   }
+
+   protected static CommonData readCommonDataFrom(ObjectInput input) throws IOException, ClassNotFoundException {
+      Object key = input.readObject();
+      Object value = input.readObject();
+      MetaParamsInternalMetadata internalMetadata = (MetaParamsInternalMetadata) input.readObject();
+      return new CommonData(key, value, internalMetadata);
+   }
+
+   protected static class CommonData {
```

Comment:
I'm not a fan of this approach. I think it hurts readability and the benefits of the `read/write` methods will soon be lost with the removal of the Externalizers